_posts/2021-09-XX-ansible-dokku.md

```diff
@@ -0,0 +1,401 @@
+  with_items: "{{ dokku_user.apps }}"
+```
+
+And finally, for each app listed we will (1) create the app, (2) give the user access to manage the app, and (3) apply the resource limits to it.
+For the moment, resource limits are global but the approach above illustrates how they could be adapted per app.
+
+`app-single.yml`: ansible task list that provisions a single app for a user
+```yaml
+---
+- name: Create app {{ dokku_app.name }}
+  dokku_app:
+    app: "{{ dokku_app.name }}"
+  when: dokku_app.name not in allowed_apps.stdout_lines
+
+- name: Let user {{ dokku_user.username }} manage app {{ dokku_app.name}}
+  command: dokku acl:add {{ dokku_app.name }} {{ dokku_user.username }}
+  when: dokku_app.name not in allowed_apps.stdout_lines
+
+- name: Set default resource limits for app {{ dokku_app.name}}
+  command: >
+    dokku resource:limit \
+      --memory {{ my_dokku_memory_limit }} \
+      --cpu {{ my_dokku_cpu_limit }} \
+      {{ dokku_app.name }}
+```
+
+The following will create the new apps and give the users access to them:
+
+```
+ansible-playbook playbook.yml --tags my_dokku_users
+```
+
+
+Finally, let's add some global configuration concerning which commands regular users are allowed to run
+```yaml
+  # see https://github.com/dokku-community/dokku-acl
+  - name: Configure ACLs, set set dokku superuser "{{ my_dokku_superuser }}"
+    become: true
+    become_user: dokku
+    lineinfile:
+      path: "~dokku/.dokkurc/acl"
+      line: "{{ item }}"
+      create: true
+    with_items:
+      - export DOKKU_ACL_ALLOW_COMMAND_LINE=1
+      - export DOKKU_SUPER_USER={{ my_dokku_superuser }}
+      # limit commands that non-superusers can run globally
+      - export DOKKU_ACL_USER_COMMANDS="help version"
+      - export DOKKU_ACL_PER_APP_COMMANDS="logs urls ps:rebuild ps:restart ps:stop ps:start git-upload-pack git-upload-archive git-receive-pack git-hook storage:list proxy:ports proxy:ports-set proxy:report enter config:bundle config:clear config:export config:get config:keys config:set config:unset"
+    tags: my_dokku_superuser
+```
+
+```
+ansible-playbook playbook.yml --tags my_dokku_superuser
+```
+
+## Conclusions
+
+That's it - you now have a dokku instance managed through ansible.
+
+See [here](https://github.com/ltalirz/ansible-playbook-dokku) for the code used in this blog post.
+Feel free to open issues on the tracker if you notice things to improve.
+
+## Further reading
+
+#### Modularization with ansible roles
+
+For the sake of brevity, in this post we have added all instructions directly to the `playbook.yml`.
+With this approach, over time the playbook can become long and hard to read.
+For increased modularity, create a new `my_dokku` ansible role in the `./roles` directory, bundle the tasks in there, and then `include_role` the role in the playbook.
+See the [ansible docs](https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html) for more information.
+
+#### Certificate management
```

Comment:
I think this section veers into more than we should for using ansible dokku. If anything, perhaps a small doc site for the plugin would be better for this "tutorial".

That or maybe the manual, non-dokku command work should be better structured as a plugin of some sort?

Reply:
My thinking here was just that if I'm someone who wants to use this article to set up a dokku instance, then certificate management is probably the main thing missing here that anyone will need to do.

I'm happy to incorporate anything written here to some other place and then link to it from here.
I'm also not sure about the wildcard thing vs one cert per app in general - what is your take on this, and is there guidance on it somewhere else?

Reply:
I almost feel like we should instead use the letsencrypt plugin to show this off instead? But doing a wildcard cert is interesting for sure.

Maybe having an `examples` folder in ansible-dokku that explains some of these more advanced this might be good? That way folks can iterate on them as time goes on (and we'd like to them from here of course).

Reply:
Ok, I'll have a look at how the letsencrypt plugin works.

Some example playbooks [are already part of the README](https://github.com/dokku/ansible-dokku#example-playbooks).
Should we just add to those or create an example folder (either solution works for me).

We can also take the example apps from there, if you think they fit better.

When I come back to this, I'll have a look at the playbooks there and update them where appropriate

Reply:
I agree that using the `dokku_letsencrypt` module seems like the way to go here, as it makes things much less complicated.
I've kept a couple of words on wildcard certificates but just as "food for thought" without code.